### PR TITLE
feat: transform media data

### LINKED ISSUE #10: Import data related to media
e.g. (titlebasics, titleakas, titleepisode, titleratings)

### PR DESCRIPTION
**Pull request by sofushn**
This PR adds a script for transforming the core data from `imdb.backup` tables into new tables.

- added: imdb language code column to `country` and ``language`
- added: integer based id named `country_id` and `language_id` to country and languages tables respective 
- added: script for transforming core imdb data into new tables, including:
  - moving all existing data tables to new schema named `original` to make it easier to distinguish between new and old tables
  - truncate data in all tables before running script to ensure it can be run multiple times
  - custom insert of country and language with correct iso codes and country/language names
  - insert of `media`, `media_genre`, `release`, `episode`, and `score`
  - generate custom `media` entries for seasons together with `release` since they don't have their own entry in the `title_basics` table
  - add reference from each `episode` to its parent `season` and reference from season to respective `series`
- changed: iso code for country and language to be based on ISO 3166-3 and ISO 649-3 respectively
- changed: `release` `title` type to `TEXT`
- fixed: `score_id` mistakenly also act as foreign key for `media`

### Missing data

the `title_akas` table contains a column named `attributes`, this seems to be additional information about the release besides in `types` column. There is currently no way this data fits into the current model, so we will have to determine a way to integrate into the current model

closes #10

### CODE REVIEW

File: src/core_data_transform.sql [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+END $$;
+
+-- Countries
+
+INSERT INTO country (code, name) 
+SELECT DISTINCT region, '' FROM original.title_akas;
+
+-- Languages
+
+INSERT INTO "language" (code, name)
```
Comment (MarLaslo) [inline src/core_data_transform.sql]: image

This probably needs a fix

Comment (sofushn) [reply src/core_data_transform.sql]: yep, should be fixed now

**Comment by MarLaslo** [comment]
Nice country codes

**Review by idax6797 — CHANGES_REQUESTED** [review]
Look super. Although I noticed that e.g. the episode_number field in the episode table was changed from NOT NULL to allow null values - is this good practice to alter?

**Comment by sofushn** [comment]
> Look super. Although I noticed that e.g. the episode_number field in the episode table was changed from NOT NULL to allow null values - is this good practice to alter?

I actually also changed the season number to allow for `NULL`. Both the episode number and season number can apparently be null 🤷 so we need to change it regardless, it should probably have been in a separate PR though 🤔 

I think the best practice would be to use an alter table statement if we weren't working on scripts that drops the entire database each time they run. However, since that is the case I would argue that its fine in this situation (open for pushback though).

**Comment by idax6797** [comment]
Thanks for the clarification! The more I think about it, the more I agree that it’s fine to allow NULL values for both episode_number and season_number. It makes sense for cases where an episode or series might be special, standalone, or unique, where assigning an episode or season number isn’t relevant.

**Comment by idax6797** [comment]
> > Look super. Although I noticed that e.g. the episode_number field in the episode table was changed from NOT NULL to allow null values - is this good practice to alter?
> 
> I actually also changed the season number to allow for `NULL`. Both the episode number and season number can apparently be null 🤷 so we need to change it regardless, it should probably have been in a separate PR though 🤔
> 
> I think the best practice would be to use an alter table statement if we weren't working on scripts that drops the entire database each time they run. However, since that is the case I would argue that its fine in this situation (open for pushback though).

**Comment by idax6797** [comment]
> > > Look super. Although I noticed that e.g. the episode_number field in the episode table was changed from NOT NULL to allow null values - is this good practice to alter?
> > 
> > 
> > I actually also changed the season number to allow for `NULL`. Both the episode number and season number can apparently be null 🤷 so we need to change it regardless, it should probably have been in a separate PR though 🤔
> > I think the best practice would be to use an alter table statement if we weren't working on scripts that drops the entire database each time they run. However, since that is the case I would argue that its fine in this situation (open for pushback though).



> > Look super. Although I noticed that e.g. the episode_number field in the episode table was changed from NOT NULL to allow null values - is this good practice to alter?
> 
> I actually also changed the season number to allow for `NULL`. Both the episode number and season number can apparently be null 🤷 so we need to change it regardless, it should probably have been in a separate PR though 🤔
> 
> I think the best practice would be to use an alter table statement if we weren't working on scripts that drops the entire database each time they run. However, since that is the case I would argue that its fine in this situation (open for pushback though).


Thanks for the clarification! The more I think about it, the more I agree that it’s fine to allow NULL values for both episode_number and season_number. It makes sense for cases where an episode or series might be special, standalone, or unique, where assigning an episode or season number isn’t relevant.

**Comment by idax6797** [comment]
> > Look super. Although I noticed that e.g. the episode_number field in the episode table was changed from NOT NULL to allow null values - is this good practice to alter?
> 
> I actually also changed the season number to allow for `NULL`. Both the episode number and season number can apparently be null 🤷 so we need to change it regardless, it should probably have been in a separate PR though 🤔
> 
> I think the best practice would be to use an alter table statement if we weren't working on scripts that drops the entire database each time they run. However, since that is the case I would argue that its fine in this situation (open for pushback though).

Thanks for the clarification! The more I think about it, the more I agree that it’s fine to allow NULL values for both episode_number and season_number. It makes sense for cases where an episode or series might be special, standalone, or unique, where assigning an episode or season number isn’t relevant.

**Review by idax6797 — APPROVED** [review]

**Pull request merged**